crates/net/network/src/budget.rs

```diff
@@ -0,0 +1,79 @@
+/// Default budget to try and drain streams.
+///
+/// Default is 1 024 iterations.
+pub const DEFAULT_BUDGET_TRY_DRAIN_STREAM: u32 = 1024;
+
+/// Default budget to try and drain `crate::Swarm`.
+///
+/// Default is 10 `crate::SwarmEvent`s.
+pub const DEFAULT_BUDGET_TRY_DRAIN_SWARM: u32 = 10;
+
+/// Default budget to try and drain pending messages from [`NetworkHandle`](crate::NetworkHandle)
+/// channel. Polling the [`TransactionsManager`](crate::transactions::TransactionsManager) future
+/// sends these types of messages.
+//
+// Default is 4 096 outgoing transaction messages.
+pub const DEFAULT_BUDGET_TRY_DRAIN_NETWORK_HANDLE_CHANNEL: u32 = 4 * 1024;
+
+/// Default budget to try and drain stream of
+/// [`NetworkTransactionEvent`](crate::transactions::NetworkTransactionEvent)s from
+/// [`NetworkManager`](crate::NetworkManager).
+///
+/// Default is 4 096 incoming transaction messages.
+pub const DEFAULT_BUDGET_TRY_DRAIN_NETWORK_TRANSACTION_EVENTS: u32 = 4 * 1024;
+
+/// Default budget to try and flush pending pool imports to pool. This number reflects the number
+/// of transactions that can be queued for import to pool in each iteration of the loop in the
+/// [`TransactionsManager`](crate::transactions::TransactionsManager) future.
+//
+// Default is 3k million pending pool imports.
+pub const DEFAULT_BUDGET_TRY_DRAIN_PENDING_POOL_IMPORTS: u32 = 3 * 100000000;
+
+/// Default budget to try and stream hashes of successfully imported transactions from the pool.
+///
+/// Default is naturally same as the number of transactions to attempt importing,
+/// [`DEFAULT_BUDGET_TRY_DRAIN_PENDING_POOL_IMPORTS`], so 3k million pool imports.
+pub const DEFAULT_BUDGET_TRY_DRAIN_POOL_IMPORTS: u32 = 3 * 100000000;
+
+/// Budget pool once.
+pub const BUDGET_ONCE: u32 = 1;
+
+/// Polls the given stream. Breaks with `true` if there maybe is more work.
+#[macro_export]
+macro_rules! poll_nested_stream_with_yield_points {
+    ($target:literal, $label:literal, $budget:ident, $poll_stream:expr, $on_ready_some:expr $(, $on_ready_none:expr;)? $(,)?) => {{
+        let mut budget: u32 = $budget;
+
+            loop {
+                match $poll_stream {
+                    Poll::Ready(Some(item)) => {
+                        let mut f = $on_ready_some;
+                        f(item);
+
+                        budget = budget.saturating_sub(1);
```

Comment:
saturating seems unnecessary

Reply:
there is never any reason not to be extra safe when it has little overhead

Reply:
but it is redundant, 
we can use an i32 for the budget and do < 0 if you're concerned about arithmetic errors, but I don't think that's necessary